### ISSUE TITLE
Compiling poppler using non-standard paths

### ISSUE THREAD

**Issue by MattFenelon** [ruby-gnome2/ruby-gnome2#733]
I'm trying to use the poppler gem on Heroku. I'm installing the `libpoppler-glib-dev` dependency using the [heroku-buildpack-apt](https://github.com/ddollar/heroku-buildpack-apt/) buildpack, which installs to a non-standard path of `/app/.apt/`. The non-standard path seems to cause problems when the poppler gem is compiled because it can't find `libpoppler-glib-dev`:

```
rbpoppler.h:29:21: fatal error: poppler.h: No such file or directory
 #include <poppler.h>
```

Is there a way to specify where the `libpoppler-glib-dev` is installed when building the gem?

Full log:
```
remote:        Installing poppler 3.0.7 with native extensions
remote:
remote:        Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
remote:
remote:        current directory: /tmp/build_85a34a6d4ee8035186889b519f8e3186/vendor/bundle/ruby/2.3.0/gems/poppler-3.0.7/ext/poppler
remote:        /tmp/build_85a34a6d4ee8035186889b519f8e3186/vendor/ruby-2.3.0/bin/ruby -r ./siteconf20160322-373-pgrlwf.rb extconf.rb
remote:        checking for --enable-debug-build option... no
remote:        checking for -Wall option to compiler... yes
remote:        checking for -Waggregate-return option to compiler... yes
remote:        checking for -Wcast-align option to compiler... yes
remote:        checking for -Wextra option to compiler... yes
remote:        checking for -Wformat=2 option to compiler... yes
remote:        checking for -Winit-self option to compiler... yes
remote:        checking for -Wlarger-than-65500 option to compiler... yes
remote:        checking for -Wmissing-declarations option to compiler... yes
remote:        checking for -Wmissing-format-attribute option to compiler... yes
remote:        checking for -Wmissing-include-dirs option to compiler... yes
remote:        checking for -Wmissing-noreturn option to compiler... yes
remote:        checking for -Wmissing-prototypes option to compiler... yes
remote:        checking for -Wnested-externs option to compiler... yes
remote:        checking for -Wold-style-definition option to compiler... yes
remote:        checking for -Wpacked option to compiler... yes
remote:        checking for -Wp,-D_FORTIFY_SOURCE=2 option to compiler... yes
remote:        checking for -Wpointer-arith option to compiler... yes
remote:        checking for -Wswitch-default option to compiler... yes
remote:        checking for -Wswitch-enum option to compiler... yes
remote:        checking for -Wundef option to compiler... yes
remote:        checking for -Wout-of-line-declaration option to compiler... no
remote:        checking for -Wunsafe-loop-optimizations option to compiler... yes
remote:        checking for -Wwrite-strings option to compiler... yes
remote:        checking for rb_define_alloc_func() in ruby.h... yes
remote:        checking for rb_block_proc() in ruby.h... yes
remote:        checking for new allocation framework... yes
remote:        checking for attribute assignment... no
remote:        checking for cairo... yes
remote:        checking for rb_cairo.h... yes
remote:        checking for Windows... no
remote:        checking for poppler-glib version (>= 0.12.0)... yes
remote:        checking for POPPLER_MAJOR_VERSION in poppler.h... no
remote:        creating rbpopplerversion.h
remote:        creating ruby-poppler.pc
remote:        creating Makefile
remote:
remote:        To see why this extension failed to compile, please check the mkmf.log which can be found here:
remote:
remote:        /tmp/build_85a34a6d4ee8035186889b519f8e3186/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/poppler-3.0.7/mkmf.log
remote:
remote:        current directory: /tmp/build_85a34a6d4ee8035186889b519f8e3186/vendor/bundle/ruby/2.3.0/gems/poppler-3.0.7/ext/poppler
remote:        make "DESTDIR=" clean
remote:
remote:        current directory: /tmp/build_85a34a6d4ee8035186889b519f8e3186/vendor/bundle/ruby/2.3.0/gems/poppler-3.0.7/ext/poppler
remote:        make "DESTDIR="
remote:        compiling rbpoppler-rectangle.c
remote:        cc1: warning: /usr/include/poppler/glib: No such file or directory [enabled by default]
remote:        cc1: warning: /usr/include/poppler: No such file or directory [enabled by default]
remote:        In file included from rbpoppler-private.h:4:0,
remote:        from rbpoppler-rectangle.c:22:
remote:        rbpoppler.h:29:21: fatal error: poppler.h: No such file or directory
remote:        #include <poppler.h>
remote:        ^
remote:        compilation terminated.
remote:        make: *** [rbpoppler-rectangle.o] Error 1
remote:
remote:        make failed, exit code 2
remote:
remote:        Gem files will remain installed in /tmp/build_85a34a6d4ee8035186889b519f8e3186/vendor/bundle/ruby/2.3.0/gems/poppler-3.0.7 for inspection.
remote:        Results logged to /tmp/build_85a34a6d4ee8035186889b519f8e3186/vendor/bundle/ruby/2.3.0/extensions/x86_64-linux/2.3.0-static/poppler-3.0.7/gem_make.out
remote:        An error occurred while installing poppler (3.0.7), and Bundler cannot continue.
remote:        Make sure that `gem install poppler -v '3.0.7'` succeeds before bundling.
```

Contents of /app/.apt:
```
/app/.apt
/app/.apt/usr
/app/.apt/usr/include
/app/.apt/usr/include/poppler
/app/.apt/usr/include/poppler/splash
/app/.apt/usr/include/poppler/goo
/app/.apt/usr/include/poppler/fofi
/app/.apt/usr/include/poppler/glib
/app/.apt/usr/lib
/app/.apt/usr/lib/x86_64-linux-gnu
/app/.apt/usr/lib/x86_64-linux-gnu/pkgconfig
/app/.apt/usr/lib/girepository-1.0
/app/.apt/usr/share
/app/.apt/usr/share/gir-1.0
/app/.apt/usr/share/doc
/app/.apt/usr/share/doc/libpoppler-private-dev
/app/.apt/usr/share/doc/libpoppler44
/app/.apt/usr/share/doc/libpoppler-dev
/app/.apt/usr/share/doc/libpoppler-glib-dev
/app/.apt/usr/share/doc/gir1.2-poppler-0.18
/app/.apt/usr/share/doc/libpoppler-glib8
```

**Comment by kou** [comment]
Please fix path in `/app/.apt/usr/lib/x86_64-linux-gnu/pkgconfig`.

**Comment by kou** [comment]
Or please try `gem install poppler -- --with-override-variables='prefix=/app/.apt/usr'`.

**Comment by MattFenelon** [comment]
Thanks @kou, I'm sort of getting there.

`gem install poppler -- --with-override-variables='prefix=/app/.apt/usr'` works, when I run it manually. The problem is that the prefix path is randomly generated, I can't configure bundler to use those args because I don't know what the path will be.

The heroku-buildpack-apt buildpack sets some environment variables for where it installs the libpoppler-glib-dev package, `CPATH`, `LD_LIBRARY_PATH`, `LIBRARY_PATH`, `INCLUDE_PATH`, `CPPPATH`, `PKG_CONFIG_PATH`. I know almost nothing about make and mkmf, do you know why the poppler makefile isn't using those variables? It's my understanding that `make` should use them by default.

**Comment by kou** [comment]
`PKG_CONFIG_PATH` is used. Your `poppler-glib.pc` in `PKG_CONFIG_PATH` returns wrong location. heroku-buildpack-apt should rewrite .pc because it changes install prefix.

**Comment by MattFenelon** [comment]
Thanks for your help @kou, you set me on the right path. The `.pc` files were wrong because heroku-buildpack-apt was downloading the apt sources, which had been built with a specific prefix in mind. I decided to compile poppler from source instead, setting the prefix for a path that makes sense in Heroku's environment (`/app/vendor/poppler`). I packaged that up as a buildpack: https://github.com/MattFenelon/heroku-buildpack-ruby-poppler.
